### PR TITLE
Check against non-nullable known primitives when rendering class blocks

### LINKED ISSUE #991: Incorrect reading of the type in codegen
Hi
When I use the new codegen version 1.9.1, I get the NoSuchMethodException error. Occurs only in classes that have boxed primitives nullable values for example `Int?`. I discovered that this has to do with the new way of creating objects. When we add, for example, `Int?` this generated adapter instead of creating a constructor for "java.lang.Integer" still uses the primitive value of `int` causing the NoSuchMethodException error.

### PR DESCRIPTION
**Pull request by ZacSweers**
`Int?` will not equal `Int` in KotlinPoet, so this was always falling through to the default `::class.java` code.

Resolves #991

### CODE REVIEW

File: kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt [hunk 1]
```diff
@@ -436,6 +436,35 @@ class DualKotlinTest(useReflection: Boolean) {
...
+
+    @Language("JSON")
+    val testInputJson = """{"oh_boy":"testing","that_is_a_test":3}"""
+    val result = adapter.fromJson(testInputJson)!!
+    assertThat(result.ohboy).isEqualTo("testing")
+    assertThat(result.thatIsATest).isEqualTo(3)
+    assertThat(result.testAgain).isEqualTo(0)
+  }
+
+  // public Testy(String var1, Integer var2, int var3, int var4, DefaultConstructorMarker var5)
```
Comment (cketti) [inline kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt]: The comment refers to `Testy`, the original class name in the bug report. The test itself uses `NullablePrimitives`.

File: kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt [hunk 2]
```diff
@@ -436,6 +436,35 @@ class DualKotlinTest(useReflection: Boolean) {
...
+  @JsonClass(generateAdapter = true)
+  data class NullablePrimitives(
```
Comment (cketti) [inline kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt]: I think this test would be easier to understand if the class contained only a single `Int?` property. Or maybe one nullable property for each primitive type.

File: kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt (hunk 1 above)
Comment (ZacSweers) [reply kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt]: Ah this was accidentally left over, will remove

File: kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt (hunk 2 above)
Comment (ZacSweers) [reply kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt]: Good call, I'll switch this out to be the different primitive types

File: kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt (hunk 1 above)
Comment (ZacSweers) [reply kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt]: 7b4e825

File: kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt (hunk 2 above)
Comment (ZacSweers) [reply kotlin/tests/src/test/kotlin/com/squareup/moshi/kotlin/DualKotlinTest.kt]: 1e22cb2

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/api/kotlintypes.kt [hunk 3]
```diff
@@ -71,7 +71,9 @@ internal fun TypeName.asTypeBlock(): CodeBlock {
       return bound.asTypeBlock()
     }
     is ClassName -> {
-      return when (this) {
+      // Check against the non-nullable version for equality, but we'll keep the nullability in
```
Comment (swankjesse) [inline kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/api/kotlintypes.kt]: Neat!

**Pull request merged**